// testcafe/lib/test-run/browser-console-messages.js
'use strict';

exports.__esModule = true;

var _getIterator2 = require('babel-runtime/core-js/get-iterator');

var _getIterator3 = _interopRequireDefault(_getIterator2);

var _assignable = require('../utils/assignable');

var _assignable2 = _interopRequireDefault(_assignable);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class BrowserConsoleMessages extends _assignable2.default {
    constructor(obj) {
        super();

        this.log = [];
        this.info = [];
        this.warn = [];
        this.error = [];

        this._assignFrom(obj);
    }

    _getAssignableProperties() {
        return [{ name: 'log' }, { name: 'info' }, { name: 'warn' }, { name: 'error' }];
    }

    concat(consoleMessages) {
        this.log = this.log.concat(consoleMessages.log);
        this.info = this.info.concat(consoleMessages.info);
        this.warn = this.warn.concat(consoleMessages.warn);
        this.error = this.error.concat(consoleMessages.error);
    }

    addMessage(type, msg) {
        this[type].push(msg);
    }

    getCopy() {
        const copy = {};
        const properties = this._getAssignableProperties();

        for (var _iterator = properties, _isArray = Array.isArray(_iterator), _i = 0, _iterator = _isArray ? _iterator : (0, _getIterator3.default)(_iterator);;) {
            var _ref;

            if (_isArray) {
                if (_i >= _iterator.length) break;
                _ref = _iterator[_i++];
            } else {
                _i = _iterator.next();
                if (_i.done) break;
                _ref = _i.value;
            }

            const property = _ref;

            copy[property.name] = this[property.name].slice();
        }return copy;
    }
}
exports.default = BrowserConsoleMessages; // -------------------------------------------------------------
// WARNING: this file is used by both the client and the server.
// Do not use any browser or node-specific API!
// -------------------------------------------------------------

module.exports = exports['default'];
