// testcafe/lib/utils/delegated-api.js
'use strict';

exports.__esModule = true;

var _defineProperty = require('babel-runtime/core-js/object/define-property');

var _defineProperty2 = _interopRequireDefault(_defineProperty);

var _getOwnPropertyNames = require('babel-runtime/core-js/object/get-own-property-names');

var _getOwnPropertyNames2 = _interopRequireDefault(_getOwnPropertyNames);

exports.getDelegatedAPIList = getDelegatedAPIList;
exports.delegateAPI = delegateAPI;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const API_IMPLEMENTATION_METHOD_RE = /^_(\S+)\$(getter|setter)?$/;

function getDelegatedAPIList(src) {
    return (0, _getOwnPropertyNames2.default)(src).map(prop => {
        const match = prop.match(API_IMPLEMENTATION_METHOD_RE);

        if (match) {
            return {
                srcProp: prop,
                apiProp: match[1],
                accessor: match[2]
            };
        }

        return null;
    }).filter(item => !!item);
}

function delegateAPI(dest, apiList, opts) {
    apiList.forEach(({ srcProp, apiProp, accessor }) => {
        const fn = function fn(...args) {
            if (opts.proxyMethod) opts.proxyMethod();

            let handler = null;

            if (opts.useCurrentCtxAsHandler) handler = this;else if (opts.getHandler) handler = opts.getHandler(apiProp, accessor);else handler = opts.handler;

            return handler[srcProp](...args);
        };

        if (accessor === 'getter') (0, _defineProperty2.default)(dest, apiProp, { get: fn, configurable: true });else if (accessor === 'setter') (0, _defineProperty2.default)(dest, apiProp, { set: fn, configurable: true });else dest[apiProp] = fn;
    });
}
