// testcafe/lib/utils/handle-errors.js
'use strict';

exports.__esModule = true;

var _values = require('babel-runtime/core-js/object/values');

var _values2 = _interopRequireDefault(_values);

exports.registerErrorHandlers = registerErrorHandlers;
exports.addRunningTest = addRunningTest;
exports.removeRunningTest = removeRunningTest;
exports.startHandlingTestErrors = startHandlingTestErrors;
exports.stopHandlingTestErrors = stopHandlingTestErrors;

var _testRun = require('../errors/test-run');

var _util = require('util');

var _util2 = _interopRequireDefault(_util);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const runningTests = {};
let handlingTestErrors = false;

function printErrorMessagesAndTerminate(...messages) {
    // eslint-disable-next-line no-console
    messages.map(msg => console.log(msg));

    // eslint-disable-next-line no-process-exit
    setTimeout(() => process.exit(1), 0);
}

function handleTestRunError(ErrorCtor, message) {
    (0, _values2.default)(runningTests).forEach(testRun => {
        testRun.addError(new ErrorCtor(message));

        removeRunningTest(testRun);
    });
}

function handleError(ErrorCtor, message) {
    if (handlingTestErrors) handleTestRunError(ErrorCtor, message);else printErrorMessagesAndTerminate(message);
}

function formatUnhandledRejectionReason(reason) {
    const reasonType = typeof reason;
    const isPrimitiveType = reasonType !== 'object' && reasonType !== 'function';

    if (isPrimitiveType) return String(reason);

    if (reason instanceof Error) return reason.stack;

    return _util2.default.inspect(reason, { depth: 2, breakLength: Infinity });
}

function formatError(ErrorCtor, error) {
    if (ErrorCtor === _testRun.UncaughtExceptionError) return error.stack;

    if (ErrorCtor === _testRun.UnhandledPromiseRejectionError) return formatUnhandledRejectionReason(error);

    return error;
}

function handleUnexpectedError(ErrorCtor, error) {
    try {
        handleError(ErrorCtor, formatError(ErrorCtor, error));
    } catch (e) {
        printErrorMessagesAndTerminate(error, e);
    }
}

function registerErrorHandlers() {
    process.on('unhandledRejection', e => handleUnexpectedError(_testRun.UnhandledPromiseRejectionError, e));
    process.on('uncaughtException', e => handleUnexpectedError(_testRun.UncaughtExceptionError, e));
}

function addRunningTest(testRun) {
    runningTests[testRun.id] = testRun;
}

function removeRunningTest(testRun) {
    if (testRun) delete runningTests[testRun.id];
}

function startHandlingTestErrors() {
    handlingTestErrors = true;
}

function stopHandlingTestErrors() {
    handlingTestErrors = false;
}
