// testcafe-hammerhead/lib/processing/dom/attributes.js
'use strict';

exports.__esModule = true;
// -------------------------------------------------------------
// WARNING: this file is used by both the client and the server.
// Do not use any browser or node-specific API!
// -------------------------------------------------------------

const URL_ATTR_TAGS = exports.URL_ATTR_TAGS = {
    href: ['a', 'link', 'image', 'area', 'base'],
    src: ['img', 'embed', 'script', 'source', 'video', 'audio', 'input', 'frame', 'iframe'],
    action: ['form'],
    formaction: ['button', 'input'],
    manifest: ['html'],
    data: ['object']
};

const URL_ATTRS = exports.URL_ATTRS = ['href', 'src', 'action', 'formaction', 'manifest', 'data'];

const TARGET_ATTR_TAGS = exports.TARGET_ATTR_TAGS = {
    target: ['a', 'form', 'area', 'base'],
    formtarget: ['input', 'button']
};

const TARGET_ATTRS = exports.TARGET_ATTRS = ['target', 'formtarget'];

const ATTRS_WITH_SPECIAL_PROXYING_LOGIC = exports.ATTRS_WITH_SPECIAL_PROXYING_LOGIC = ['sandbox', 'autocomplete', 'target', 'formtarget', 'style'];
