// testcafe/lib/utils/temp-directory/index.js
'use strict';

exports.__esModule = true;

var _values = require('babel-runtime/core-js/object/values');

var _values2 = _interopRequireDefault(_values);

var _getIterator2 = require('babel-runtime/core-js/get-iterator');

var _getIterator3 = _interopRequireDefault(_getIterator2);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _debug = require('debug');

var _debug2 = _interopRequireDefault(_debug);

var _os = require('os');

var _os2 = _interopRequireDefault(_os);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _asyncExitHook = require('async-exit-hook');

var _asyncExitHook2 = _interopRequireDefault(_asyncExitHook);

var _tmp = require('tmp');

var _tmp2 = _interopRequireDefault(_tmp);

var _makeDir = require('make-dir');

var _makeDir2 = _interopRequireDefault(_makeDir);

var _lockfile = require('./lockfile');

var _lockfile2 = _interopRequireDefault(_lockfile);

var _cleanupProcess = require('./cleanup-process');

var _cleanupProcess2 = _interopRequireDefault(_cleanupProcess);

var _promisifiedFunctions = require('../../utils/promisified-functions');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// NOTE: mutable for testing purposes
const TESTCAFE_TMP_DIRS_ROOT = _path2.default.join(_os2.default.tmpdir(), 'testcafe');
const DEFAULT_NAME_PREFIX = 'tmp';
const USED_TEMP_DIRS = {};
const DEBUG_LOGGER = (0, _debug2.default)('testcafe:utils:temp-directory');

class TempDirectory {
    constructor(namePrefix) {
        this.namePrefix = namePrefix || DEFAULT_NAME_PREFIX;

        this.path = '';
        this.lockFile = null;
    }

    _getTmpDirsList() {
        var _this = this;

        return (0, _asyncToGenerator3.default)(function* () {
            const tmpDirNames = yield (0, _promisifiedFunctions.readDir)(TempDirectory.TEMP_DIRECTORIES_ROOT);

            return tmpDirNames.filter(function (tmpDir) {
                return !USED_TEMP_DIRS[tmpDir];
            }).filter(function (tmpDir) {
                return _path2.default.basename(tmpDir).startsWith(_this.namePrefix);
            });
        })();
    }

    _findFreeTmpDir(tmpDirNames) {
        var _this2 = this;

        return (0, _asyncToGenerator3.default)(function* () {
            for (var _iterator = tmpDirNames, _isArray = Array.isArray(_iterator), _i = 0, _iterator = _isArray ? _iterator : (0, _getIterator3.default)(_iterator);;) {
                var _ref;

                if (_isArray) {
                    if (_i >= _iterator.length) break;
                    _ref = _iterator[_i++];
                } else {
                    _i = _iterator.next();
                    if (_i.done) break;
                    _ref = _i.value;
                }

                const tmpDirName = _ref;

                const tmpDirPath = _path2.default.join(TempDirectory.TEMP_DIRECTORIES_ROOT, tmpDirName);

                const lockFile = new _lockfile2.default(tmpDirPath);

                if (lockFile.init()) {
                    _this2.path = tmpDirPath;
                    _this2.lockFile = lockFile;

                    return true;
                }
            }

            return false;
        })();
    }

    _createNewTmpDir() {
        var _this3 = this;

        return (0, _asyncToGenerator3.default)(function* () {
            _this3.path = _tmp2.default.tmpNameSync({ dir: TempDirectory.TEMP_DIRECTORIES_ROOT, prefix: _this3.namePrefix + '-' });

            yield (0, _makeDir2.default)(_this3.path);

            _this3.lockFile = new _lockfile2.default(_this3.path);

            _this3.lockFile.init();
        })();
    }

    _disposeSync() {
        if (!USED_TEMP_DIRS[this.path]) return;

        this.lockFile.dispose();

        delete USED_TEMP_DIRS[this.path];
    }

    static createDirectory(prefix) {
        return (0, _asyncToGenerator3.default)(function* () {
            const tmpDir = new TempDirectory(prefix);

            yield tmpDir.init();

            return tmpDir;
        })();
    }

    static disposeDirectoriesSync() {
        (0, _values2.default)(USED_TEMP_DIRS).forEach(tmpDir => tmpDir._disposeSync());
    }

    init() {
        var _this4 = this;

        return (0, _asyncToGenerator3.default)(function* () {
            yield (0, _makeDir2.default)(TempDirectory.TEMP_DIRECTORIES_ROOT);

            const tmpDirNames = yield _this4._getTmpDirsList(_this4.namePrefix);

            DEBUG_LOGGER('Found temp directories:', tmpDirNames);

            const existingTmpDirFound = yield _this4._findFreeTmpDir(tmpDirNames);

            if (!existingTmpDirFound) yield _this4._createNewTmpDir();

            DEBUG_LOGGER('Temp directory path: ', _this4.path);

            yield _cleanupProcess2.default.init();
            yield _cleanupProcess2.default.addDirectory(_this4.path);

            USED_TEMP_DIRS[_this4.path] = _this4;
        })();
    }

    dispose() {
        var _this5 = this;

        return (0, _asyncToGenerator3.default)(function* () {
            if (!USED_TEMP_DIRS[_this5.path]) return;

            _this5.lockFile.dispose();

            yield _cleanupProcess2.default.removeDirectory(_this5.path);

            delete USED_TEMP_DIRS[_this5.path];
        })();
    }
}

exports.default = TempDirectory; // NOTE: exposed for testing purposes

TempDirectory.TEMP_DIRECTORIES_ROOT = TESTCAFE_TMP_DIRS_ROOT;

(0, _asyncExitHook2.default)(TempDirectory.disposeDirectoriesSync);
module.exports = exports['default'];
