// @storybook/addon-info/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.setDefaults = setDefaults;
Object.defineProperty(exports, "Story", {
  enumerable: true,
  get: function get() {
    return _Story.default;
  }
});
exports.withInfo = void 0;

require("core-js/modules/es6.array.for-each");

require("core-js/modules/es6.array.filter");

require("core-js/modules/web.dom.iterable");

require("core-js/modules/es6.array.iterator");

require("core-js/modules/es6.object.keys");

require("core-js/modules/es6.object.define-property");

require("core-js/modules/es6.object.assign");

var _react = _interopRequireDefault(require("react"));

var _nestedObjectAssign = _interopRequireDefault(require("nested-object-assign"));

var _utilDeprecate = _interopRequireDefault(require("util-deprecate"));

var _addons = require("@storybook/addons");

var _clientLogger = require("@storybook/client-logger");

var _Story = _interopRequireDefault(require("./components/Story"));

var _PropTable = _interopRequireDefault(require("./components/PropTable"));

var _makeTableComponent = _interopRequireDefault(require("./components/makeTableComponent"));

var _markdown = require("./components/markdown");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var defaultOptions = {
  inline: false,
  header: true,
  source: true,
  propTables: [],
  TableComponent: _PropTable.default,
  maxPropsIntoLine: 3,
  maxPropObjectKeys: 3,
  maxPropArrayLength: 3,
  maxPropStringLength: 50
};
var defaultComponents = {
  h1: _markdown.H1,
  h2: _markdown.H2,
  h3: _markdown.H3,
  h4: _markdown.H4,
  h5: _markdown.H5,
  h6: _markdown.H6,
  code: _markdown.Code,
  p: _markdown.P,
  a: _markdown.A,
  li: _markdown.LI,
  ul: _markdown.UL
};
var hasWarned = false;

function addInfo(storyFn, context, infoOptions) {
  var options = _objectSpread({}, defaultOptions, infoOptions); // props.propTables can only be either an array of components or null
  // propTables option is allowed to be set to 'false' (a boolean)
  // if the option is false, replace it with null to avoid react warnings


  if (!options.propTables) {
    options.propTables = null;
  }

  var components = _objectSpread({}, defaultComponents);

  if (options && options.components) {
    Object.assign(components, options.components);
  }

  if (options && options.marksyConf) {
    if (!hasWarned) {
      _clientLogger.logger.warn('@storybook/addon-info: "marksyConf" option has been renamed to "components"');

      hasWarned = true;
    }

    Object.assign(components, options.marksyConf);
  }

  var props = {
    info: options.text,
    context: context,
    showInline: Boolean(options.inline),
    showHeader: Boolean(options.header),
    showSource: Boolean(options.source),
    styles: typeof options.styles === 'function' ? options.styles : function (s) {
      return (0, _nestedObjectAssign.default)({}, s, options.styles);
    },
    propTables: options.propTables,
    propTablesExclude: options.propTablesExclude,
    PropTable: (0, _makeTableComponent.default)(options.TableComponent),
    components: components,
    maxPropObjectKeys: options.maxPropObjectKeys,
    maxPropArrayLength: options.maxPropArrayLength,
    maxPropsIntoLine: options.maxPropsIntoLine,
    maxPropStringLength: options.maxPropStringLength,
    excludedPropTypes: options.excludedPropTypes
  };
  return _react.default.createElement(_Story.default, props, storyFn(context));
}

var withInfo = (0, _addons.makeDecorator)({
  name: 'withInfo',
  parameterName: 'info',
  allowDeprecatedUsage: true,
  wrapper: function wrapper(getStory, context, _ref) {
    var options = _ref.options,
        parameters = _ref.parameters;
    var storyOptions = parameters || options;
    var infoOptions = typeof storyOptions === 'string' ? {
      text: storyOptions
    } : storyOptions;
    var mergedOptions = typeof infoOptions === 'string' ? infoOptions : _objectSpread({}, options, infoOptions);
    return addInfo(getStory, context, mergedOptions);
  }
});
exports.withInfo = withInfo;

function setDefaults(newDefaults) {
  return (0, _utilDeprecate.default)(function () {
    return Object.assign(defaultOptions, newDefaults);
  }, 'setDefaults is deprecated. Instead, you can pass options into withInfo(options) directly, or use the info parameter.')();
}
