// @storybook/react/dist/server/framework-preset-cra.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.webpackFinal = webpackFinal;
exports.babelDefault = babelDefault;

var _nodeLogger = require("@storybook/node-logger");

var _craConfig = require("./cra-config");

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function webpackFinal(config, {
  configDir
}) {
  if (!(0, _craConfig.isReactScriptsInstalled)()) {
    _nodeLogger.logger.info('=> Using base config because react-scripts is not installed.');

    return config;
  }

  _nodeLogger.logger.info('=> Loading create-react-app config.');

  return (0, _craConfig.applyCRAWebpackConfig)(config, configDir);
}

function babelDefault(config) {
  if (!(0, _craConfig.isReactScriptsInstalled)()) {
    return config;
  }

  return _objectSpread({}, config, {
    presets: [require.resolve('babel-preset-react-app')],
    plugins: [[require.resolve('babel-plugin-named-asset-import'), {
      loaderMap: {
        svg: {
          ReactComponent: '@svgr/webpack?-prettier,-svgo![path]'
        }
      }
    }]]
  });
}
