// testcafe/lib/compiler/test-file/formats/coffeescript/compiler.js
'use strict';

exports.__esModule = true;

var _coffeescript = require('coffeescript');

var _coffeescript2 = _interopRequireDefault(_coffeescript);

var _loadBabelLibs2 = require('../../../load-babel-libs');

var _loadBabelLibs3 = _interopRequireDefault(_loadBabelLibs2);

var _compiler = require('../es-next/compiler.js');

var _compiler2 = _interopRequireDefault(_compiler);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const FIXTURE_RE = /(^|;|\s+)fixture\s*(\.|\(|'|")/;
const TEST_RE = /(^|;|\s+)test\s*/;

class CoffeeScriptTestFileCompiler extends _compiler2.default {
    _hasTests(code) {
        return FIXTURE_RE.test(code) && TEST_RE.test(code);
    }

    _compileCode(code, filename) {
        if (this.cache[filename]) return this.cache[filename];

        const transpiled = _coffeescript2.default.compile(code, {
            filename,
            bare: true,
            sourceMap: true,
            inlineMap: true,
            header: false
        });

        var _loadBabelLibs = (0, _loadBabelLibs3.default)();

        const babel = _loadBabelLibs.babel;

        const babelOptions = _compiler2.default.getBabelOptions(filename, code);
        const compiled = babel.transform(transpiled.js, babelOptions);

        this.cache[filename] = compiled.code;

        return compiled.code;
    }

    _getRequireCompilers() {
        return { '.coffee': (code, filename) => this._compileCode(code, filename) };
    }

    getSupportedExtension() {
        return '.coffee';
    }
}
exports.default = CoffeeScriptTestFileCompiler;
module.exports = exports['default'];
