// testcafe/lib/compiler/test-file/formats/coffeescript/get-test-list.js
'use strict';

exports.__esModule = true;
exports.getCoffeeScriptTestListFromCode = exports.getCoffeeScriptTestList = exports.CoffeeScriptTestFileParser = undefined;

var _coffeescript = require('coffeescript');

var _coffeescript2 = _interopRequireDefault(_coffeescript);

var _babelCore = require('babel-core');

var _compiler = require('../es-next/compiler.js');

var _compiler2 = _interopRequireDefault(_compiler);

var _getTestList = require('../es-next/get-test-list');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class CoffeeScriptTestFileParser extends _getTestList.EsNextTestFileParser {
    parse(code) {
        const babelOptions = _compiler2.default.getBabelOptions(null, code);

        delete babelOptions.filename;
        babelOptions.ast = true;

        code = _coffeescript2.default.compile(code, {
            bare: true,
            sourceMap: false,
            inlineMap: false,
            header: false
        });

        const ast = (0, _babelCore.transform)(code, babelOptions).ast;

        return this.analyze(ast.program.body);
    }
}

exports.CoffeeScriptTestFileParser = CoffeeScriptTestFileParser;
const parser = new CoffeeScriptTestFileParser();

const getCoffeeScriptTestList = exports.getCoffeeScriptTestList = parser.getTestList.bind(parser);
const getCoffeeScriptTestListFromCode = exports.getCoffeeScriptTestListFromCode = parser.getTestListFromCode.bind(parser);
