// testcafe/lib/api/structure/test.js
'use strict';

exports.__esModule = true;

var _from = require('babel-runtime/core-js/array/from');

var _from2 = _interopRequireDefault(_from);

var _testingUnit = require('./testing-unit');

var _testingUnit2 = _interopRequireDefault(_testingUnit);

var _typeAssertions = require('../../errors/runtime/type-assertions');

var _wrapTestFunction = require('../wrap-test-function');

var _wrapTestFunction2 = _interopRequireDefault(_wrapTestFunction);

var _assertType = require('../request-hooks/assert-type');

var _assertType2 = _interopRequireDefault(_assertType);

var _lodash = require('lodash');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class Test extends _testingUnit2.default {
    constructor(testFile) {
        super(testFile, 'test');

        this.fixture = testFile.currentFixture;

        this.fn = null;
        this.beforeFn = null;
        this.afterFn = null;
        this.requestHooks = [];

        return this.apiOrigin;
    }

    _add(name, fn) {
        (0, _typeAssertions.assertType)(_typeAssertions.is.string, 'apiOrigin', 'The test name', name);
        (0, _typeAssertions.assertType)(_typeAssertions.is.function, 'apiOrigin', 'The test body', fn);
        (0, _typeAssertions.assertType)(_typeAssertions.is.nonNullObject, 'apiOrigin', `The fixture of '${name}' test`, this.fixture);

        this.name = name;
        this.fn = (0, _wrapTestFunction2.default)(fn);
        this.requestHooks = (0, _lodash.union)(this.requestHooks, (0, _from2.default)(this.fixture.requestHooks));

        if (this.testFile.collectedTests.indexOf(this) < 0) this.testFile.collectedTests.push(this);

        return this.apiOrigin;
    }

    _before$(fn) {
        (0, _typeAssertions.assertType)(_typeAssertions.is.function, 'before', 'test.before hook', fn);

        this.beforeFn = (0, _wrapTestFunction2.default)(fn);

        return this.apiOrigin;
    }

    _after$(fn) {
        (0, _typeAssertions.assertType)(_typeAssertions.is.function, 'after', 'test.after hook', fn);

        this.afterFn = (0, _wrapTestFunction2.default)(fn);

        return this.apiOrigin;
    }

    _requestHooks$(...hooks) {
        hooks = (0, _lodash.flattenDeep)(hooks);

        (0, _assertType2.default)(hooks);

        this.requestHooks = (0, _lodash.union)(this.requestHooks, hooks);

        return this.apiOrigin;
    }
}

exports.default = Test;
_testingUnit2.default._makeAPIListForChildClass(Test);
module.exports = exports['default'];
