// testcafe/lib/compiler/test-file/formats/es-next/compiler.js
'use strict';

exports.__esModule = true;

var _loadBabelLibs3 = require('../../../load-babel-libs');

var _loadBabelLibs4 = _interopRequireDefault(_loadBabelLibs3);

var _apiBased = require('../../api-based');

var _apiBased2 = _interopRequireDefault(_apiBased);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const BABEL_RUNTIME_RE = /^babel-runtime(\\|\/|$)/;
const FLOW_MARKER_RE = /^\s*\/\/\s*@flow\s*\n|^\s*\/\*\s*@flow\s*\*\//;

class ESNextTestFileCompiler extends _apiBased2.default {
    static getBabelOptions(filename, code) {
        var _loadBabelLibs = (0, _loadBabelLibs4.default)();

        const presetStage2 = _loadBabelLibs.presetStage2,
              presetFlow = _loadBabelLibs.presetFlow,
              transformRuntime = _loadBabelLibs.transformRuntime,
              transformClassProperties = _loadBabelLibs.transformClassProperties,
              presetEnv = _loadBabelLibs.presetEnv;

        // NOTE: passPrePreset and complex presets is a workaround for https://github.com/babel/babel/issues/2877
        // Fixes https://github.com/DevExpress/testcafe/issues/969

        return {
            passPerPreset: true,
            presets: [{
                passPerPreset: false,
                presets: [{ plugins: [transformRuntime] }, presetStage2, presetEnv]
            }, FLOW_MARKER_RE.test(code) ? {
                passPerPreset: false,
                presets: [{ plugins: [transformClassProperties] }, presetFlow]
            } : {}],
            filename: filename,
            retainLines: true,
            sourceMaps: 'inline',
            ast: false,
            babelrc: false,
            highlightCode: false,

            resolveModuleSource: source => {
                if (source === 'testcafe') return _apiBased2.default.EXPORTABLE_LIB_PATH;

                if (BABEL_RUNTIME_RE.test(source)) {
                    try {
                        return require.resolve(source);
                    } catch (err) {
                        return source;
                    }
                }

                return source;
            }
        };
    }

    _compileCode(code, filename) {
        var _loadBabelLibs2 = (0, _loadBabelLibs4.default)();

        const babel = _loadBabelLibs2.babel;


        if (this.cache[filename]) return this.cache[filename];

        const opts = ESNextTestFileCompiler.getBabelOptions(filename, code);
        const compiled = babel.transform(code, opts);

        this.cache[filename] = compiled.code;

        return compiled.code;
    }

    _getRequireCompilers() {
        return { '.js': (code, filename) => this._compileCode(code, filename) };
    }

    getSupportedExtension() {
        return '.js';
    }
}
exports.default = ESNextTestFileCompiler;
module.exports = exports['default'];
