// @storybook/addon-storyshots/dist/frameworks/angular/loader.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _hasDependency = _interopRequireDefault(require("../hasDependency"));

var _configure = _interopRequireDefault(require("../configure"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function setupAngularJestPreset() {
  // Needed to prevent "Zone.js has detected that ZoneAwarePromise `(window|global).Promise` has been overwritten."
  require.requireActual('core-js/modules/es6.promise'); // Angular + Jest + Storyshots = Crazy Shit:
  // We need to require 'jest-preset-angular/setupJest' before any storybook code
  // is running inside jest -  one of the things that `jest-preset-angular/setupJest` does is
  // extending the `window.Reflect` with all the needed metadata functions, that are required
  // for emission of the TS decorations like 'design:paramtypes'


  require.requireActual('jest-preset-angular/setupJest');
}

function test(options) {
  return options.framework === 'angular' || !options.framework && (0, _hasDependency.default)('@storybook/angular');
}

function load(options) {
  setupAngularJestPreset();
  const {
    configPath,
    config
  } = options;

  const storybook = require.requireActual('@storybook/angular');

  (0, _configure.default)({
    configPath,
    config,
    storybook
  });
  return {
    framework: 'angular',
    renderTree: require.requireActual('./renderTree').default,
    renderShallowTree: () => {
      throw new Error('Shallow renderer is not supported for angular');
    },
    storybook
  };
}

var _default = {
  load,
  test
};
exports.default = _default;
