// @storybook/addon-storyshots/dist/frameworks/angular/renderTree.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _AngularSnapshotSerializer = _interopRequireDefault(require("jest-preset-angular/AngularSnapshotSerializer"));

var _HTMLCommentSerializer = _interopRequireDefault(require("jest-preset-angular/HTMLCommentSerializer"));

var _testing = require("@angular/core/testing");

var _testing2 = require("@angular/platform-browser-dynamic/testing");

var _core = require("@angular/core");

var _jestSpecificSnapshot = require("jest-specific-snapshot");

var _helpers = require("./helpers");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// eslint-disable-next-line import/no-extraneous-dependencies
// eslint-disable-next-line import/no-extraneous-dependencies
// eslint-disable-next-line import/no-extraneous-dependencies
// eslint-disable-next-line import/no-extraneous-dependencies
// eslint-disable-next-line import/no-extraneous-dependencies
(0, _jestSpecificSnapshot.addSerializer)(_HTMLCommentSerializer.default);
(0, _jestSpecificSnapshot.addSerializer)(_AngularSnapshotSerializer.default);

function getRenderedTree(story, context) {
  const currentStory = story.render(context);
  const {
    moduleMeta,
    AppComponent
  } = (0, _helpers.initModuleData)(currentStory);

  _testing.TestBed.configureTestingModule({
    imports: [...moduleMeta.imports],
    declarations: [...moduleMeta.declarations],
    providers: [...moduleMeta.providers],
    schemas: [_core.NO_ERRORS_SCHEMA, ...moduleMeta.schemas],
    bootstrap: [...moduleMeta.bootstrap]
  });

  _testing.TestBed.overrideModule(_testing2.BrowserDynamicTestingModule, {
    set: {
      entryComponents: [...moduleMeta.entryComponents]
    }
  });

  return _testing.TestBed.compileComponents().then(() => {
    const tree = _testing.TestBed.createComponent(AppComponent);

    tree.detectChanges();
    return tree;
  });
}

var _default = getRenderedTree;
exports.default = _default;
