// testcafe/lib/client-functions/selectors/add-api.js
'use strict';

exports.__esModule = true;

var _defineProperty = require('babel-runtime/core-js/object/define-property');

var _defineProperty2 = _interopRequireDefault(_defineProperty);

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _keys = require('babel-runtime/core-js/object/keys');

var _keys2 = _interopRequireDefault(_keys);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

let getSnapshot = (() => {
    var _ref = (0, _asyncToGenerator3.default)(function* (getSelector, callsite, SelectorBuilder) {
        let node = null;
        const selector = new SelectorBuilder(getSelector(), { needError: true }, { instantiation: 'Selector' }).getFunction();

        try {
            node = yield selector();
        } catch (err) {
            err.callsite = callsite;
            throw err;
        }

        return node;
    });

    return function getSnapshot(_x, _x2, _x3) {
        return _ref.apply(this, arguments);
    };
})();

exports.addCustomMethods = addCustomMethods;
exports.addAPI = addAPI;

var _lodash = require('lodash');

var _builderSymbol = require('../builder-symbol');

var _builderSymbol2 = _interopRequireDefault(_builderSymbol);

var _snapshotProperties = require('./snapshot-properties');

var _getCallsite = require('../../errors/get-callsite');

var _clientFunctionBuilder = require('../client-function-builder');

var _clientFunctionBuilder2 = _interopRequireDefault(_clientFunctionBuilder);

var _reExecutablePromise = require('../../utils/re-executable-promise');

var _reExecutablePromise2 = _interopRequireDefault(_reExecutablePromise);

var _typeAssertions = require('../../errors/runtime/type-assertions');

var _makeRegExp = require('../../utils/make-reg-exp');

var _makeRegExp2 = _interopRequireDefault(_makeRegExp);

var _selectorTextFilter = require('./selector-text-filter');

var _selectorTextFilter2 = _interopRequireDefault(_selectorTextFilter);

var _selectorAttributeFilter = require('./selector-attribute-filter');

var _selectorAttributeFilter2 = _interopRequireDefault(_selectorAttributeFilter);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const filterNodes = new _clientFunctionBuilder2.default((nodes, filter, querySelectorRoot, originNode, ...filterArgs) => {
    if (typeof filter === 'number') {
        const matchingNode = filter < 0 ? nodes[nodes.length + filter] : nodes[filter];

        return matchingNode ? [matchingNode] : [];
    }

    const result = [];

    if (typeof filter === 'string') {
        // NOTE: we can search for elements only in document or element.
        if (querySelectorRoot.nodeType !== 1 && querySelectorRoot.nodeType !== 9) return null;

        const matching = querySelectorRoot.querySelectorAll(filter);
        const matchingArr = [];

        for (let i = 0; i < matching.length; i++) matchingArr.push(matching[i]);

        filter = node => matchingArr.indexOf(node) > -1;
    }

    if (typeof filter === 'function') {
        for (let j = 0; j < nodes.length; j++) {
            if (filter(nodes[j], j, originNode, ...filterArgs)) result.push(nodes[j]);
        }
    }

    return result;
}).getFunction();

const expandSelectorResults = new _clientFunctionBuilder2.default((selector, populateDerivativeNodes) => {
    const nodes = selector();

    if (!nodes.length) return null;

    const result = [];

    for (let i = 0; i < nodes.length; i++) {
        const derivativeNodes = populateDerivativeNodes(nodes[i]);

        if (derivativeNodes) {
            for (let j = 0; j < derivativeNodes.length; j++) {
                if (result.indexOf(derivativeNodes[j]) < 0) result.push(derivativeNodes[j]);
            }
        }
    }

    return result;
}).getFunction();

function assertAddCustomDOMPropertiesOptions(properties) {
    (0, _typeAssertions.assertType)(_typeAssertions.is.nonNullObject, 'addCustomDOMProperties', '"addCustomDOMProperties" option', properties);

    (0, _keys2.default)(properties).forEach(prop => {
        (0, _typeAssertions.assertType)(_typeAssertions.is.function, 'addCustomDOMProperties', `Custom DOM properties method '${prop}'`, properties[prop]);
    });
}

function assertAddCustomMethods(properties, opts) {
    (0, _typeAssertions.assertType)(_typeAssertions.is.nonNullObject, 'addCustomMethods', '"addCustomMethods" option', properties);

    if (opts !== void 0) (0, _typeAssertions.assertType)(_typeAssertions.is.nonNullObject, 'addCustomMethods', '"addCustomMethods" option', opts);

    (0, _keys2.default)(properties).forEach(prop => {
        (0, _typeAssertions.assertType)(_typeAssertions.is.function, 'addCustomMethods', `Custom method '${prop}'`, properties[prop]);
    });
}

function prepareApiFnArgs(fnName, ...args) {
    args = args.map(arg => {
        if (typeof arg === 'string') return `'${arg}'`;
        if (typeof arg === 'function') return '[function]';
        return arg;
    });
    args = args.join(', ');

    return `.${fnName}(${args})`;
}

function getDerivativeSelectorArgs(options, selectorFn, apiFn, filter, additionalDependencies) {
    return (0, _assign2.default)({}, options, { selectorFn, apiFn, filter, additionalDependencies });
}

function addSnapshotProperties(obj, getSelector, SelectorBuilder, properties) {
    properties.forEach(prop => {
        (0, _defineProperty2.default)(obj, prop, {
            get: () => {
                const callsite = (0, _getCallsite.getCallsiteForMethod)('get');

                return _reExecutablePromise2.default.fromFn((0, _asyncToGenerator3.default)(function* () {
                    const snapshot = yield getSnapshot(getSelector, callsite, SelectorBuilder);

                    return snapshot[prop];
                }));
            }
        });
    });
}

function addCustomMethods(obj, getSelector, SelectorBuilder, customMethods) {
    const customMethodProps = customMethods ? (0, _keys2.default)(customMethods) : [];

    customMethodProps.forEach(prop => {
        var _customMethods$prop = customMethods[prop],
            _customMethods$prop$r = _customMethods$prop.returnDOMNodes;
        const returnDOMNodes = _customMethods$prop$r === undefined ? false : _customMethods$prop$r,
              method = _customMethods$prop.method;


        const dependencies = {
            customMethod: method,
            selector: getSelector()
        };

        const callsiteNames = { instantiation: prop };

        if (returnDOMNodes) {
            obj[prop] = (...args) => {
                const selectorFn = () => {
                    /* eslint-disable no-undef */
                    const nodes = selector();

                    return customMethod.apply(customMethod, [nodes].concat(args));
                    /* eslint-enable no-undef */
                };

                const apiFn = prepareApiFnArgs(prop, ...args);
                const filter = () => true;

                const additionalDependencies = {
                    args,
                    customMethod: method
                };

                return createDerivativeSelectorWithFilter({ getSelector, SelectorBuilder, selectorFn, apiFn, filter, additionalDependencies });
            };
        } else {
            obj[prop] = new _clientFunctionBuilder2.default((...args) => {
                /* eslint-disable no-undef */
                const node = selector();

                return customMethod.apply(customMethod, [node].concat(args));
                /* eslint-enable no-undef */
            }, { dependencies }, callsiteNames).getFunction();
        }
    });
}

function addSnapshotPropertyShorthands({ obj, getSelector, SelectorBuilder, customDOMProperties, customMethods }) {
    let properties = _snapshotProperties.SNAPSHOT_PROPERTIES;

    if (customDOMProperties) properties = properties.concat((0, _keys2.default)(customDOMProperties));

    addSnapshotProperties(obj, getSelector, SelectorBuilder, properties);
    addCustomMethods(obj, getSelector, SelectorBuilder, customMethods);

    obj.getStyleProperty = prop => {
        const callsite = (0, _getCallsite.getCallsiteForMethod)('getStyleProperty');

        return _reExecutablePromise2.default.fromFn((0, _asyncToGenerator3.default)(function* () {
            const snapshot = yield getSnapshot(getSelector, callsite, SelectorBuilder);

            return snapshot.style ? snapshot.style[prop] : void 0;
        }));
    };

    obj.getAttribute = attrName => {
        const callsite = (0, _getCallsite.getCallsiteForMethod)('getAttribute');

        return _reExecutablePromise2.default.fromFn((0, _asyncToGenerator3.default)(function* () {
            const snapshot = yield getSnapshot(getSelector, callsite, SelectorBuilder);

            return snapshot.attributes ? snapshot.attributes[attrName] : void 0;
        }));
    };

    obj.hasAttribute = attrName => {
        const callsite = (0, _getCallsite.getCallsiteForMethod)('hasAttribute');

        return _reExecutablePromise2.default.fromFn((0, _asyncToGenerator3.default)(function* () {
            const snapshot = yield getSnapshot(getSelector, callsite, SelectorBuilder);

            return snapshot.attributes ? snapshot.attributes.hasOwnProperty(attrName) : false;
        }));
    };

    obj.getBoundingClientRectProperty = prop => {
        const callsite = (0, _getCallsite.getCallsiteForMethod)('getBoundingClientRectProperty');

        return _reExecutablePromise2.default.fromFn((0, _asyncToGenerator3.default)(function* () {
            const snapshot = yield getSnapshot(getSelector, callsite, SelectorBuilder);

            return snapshot.boundingClientRect ? snapshot.boundingClientRect[prop] : void 0;
        }));
    };

    obj.hasClass = name => {
        const callsite = (0, _getCallsite.getCallsiteForMethod)('hasClass');

        return _reExecutablePromise2.default.fromFn((0, _asyncToGenerator3.default)(function* () {
            const snapshot = yield getSnapshot(getSelector, callsite, SelectorBuilder);

            return snapshot.classNames ? snapshot.classNames.indexOf(name) > -1 : false;
        }));
    };
}

function createCounter(getSelector, SelectorBuilder) {
    const builder = new SelectorBuilder(getSelector(), { counterMode: true }, { instantiation: 'Selector' });
    const counter = builder.getFunction();
    const callsite = (0, _getCallsite.getCallsiteForMethod)('get');

    return (0, _asyncToGenerator3.default)(function* () {
        try {
            return yield counter();
        } catch (err) {
            err.callsite = callsite;
            throw err;
        }
    });
}

function addCounterProperties({ obj, getSelector, SelectorBuilder }) {
    Object.defineProperty(obj, 'count', {
        get: () => {
            const counter = createCounter(getSelector, SelectorBuilder);

            return _reExecutablePromise2.default.fromFn(() => counter());
        }
    });

    Object.defineProperty(obj, 'exists', {
        get: () => {
            const counter = createCounter(getSelector, SelectorBuilder);

            return _reExecutablePromise2.default.fromFn((0, _asyncToGenerator3.default)(function* () {
                return (yield counter()) > 0;
            }));
        }
    });
}

function convertFilterToClientFunctionIfNecessary(callsiteName, filter, dependencies) {
    if (typeof filter === 'function') {
        const builder = filter[_builderSymbol2.default];
        const fn = builder ? builder.fn : filter;
        const options = builder ? (0, _lodash.assign)({}, builder.options, { dependencies }) : { dependencies };

        return new _clientFunctionBuilder2.default(fn, options, { instantiation: callsiteName }).getFunction();
    }

    return filter;
}

function createDerivativeSelectorWithFilter({ getSelector, SelectorBuilder, selectorFn, apiFn, filter, additionalDependencies }) {
    const collectionModeSelectorBuilder = new SelectorBuilder(getSelector(), { collectionMode: true });
    const customDOMProperties = collectionModeSelectorBuilder.options.customDOMProperties;
    const customMethods = collectionModeSelectorBuilder.options.customMethods;

    let dependencies = {
        selector: collectionModeSelectorBuilder.getFunction(),
        filter: filter,
        filterNodes: filterNodes
    };

    var _collectionModeSelect = collectionModeSelectorBuilder.options;
    const boundTestRun = _collectionModeSelect.boundTestRun,
          timeout = _collectionModeSelect.timeout,
          visibilityCheck = _collectionModeSelect.visibilityCheck,
          apiFnChain = _collectionModeSelect.apiFnChain;


    dependencies = (0, _lodash.assign)(dependencies, additionalDependencies);

    const builder = new SelectorBuilder(selectorFn, {
        dependencies,
        customDOMProperties,
        customMethods,
        boundTestRun,
        timeout,
        visibilityCheck,
        apiFnChain,
        apiFn
    }, { instantiation: 'Selector' });

    return builder.getFunction();
}

const filterByText = convertFilterToClientFunctionIfNecessary('filter', _selectorTextFilter2.default);
const filterByAttr = convertFilterToClientFunctionIfNecessary('filter', _selectorAttributeFilter2.default);

function ensureRegExpContext(str) {
    // NOTE: if a regexp is created in a separate context (via the 'vm' module) we
    // should wrap it with new RegExp() to make the `instanceof RegExp` check successful.
    if (typeof str !== 'string' && !(str instanceof RegExp)) return new RegExp(str);

    return str;
}

function addFilterMethods(options) {
    const obj = options.obj,
          getSelector = options.getSelector,
          SelectorBuilder = options.SelectorBuilder;


    obj.nth = index => {
        (0, _typeAssertions.assertType)(_typeAssertions.is.number, 'nth', '"index" argument', index);

        const apiFn = prepareApiFnArgs('nth', index);
        const builder = new SelectorBuilder(getSelector(), { index, apiFn }, { instantiation: 'Selector' });

        return builder.getFunction();
    };

    obj.withText = text => {
        (0, _typeAssertions.assertType)([_typeAssertions.is.string, _typeAssertions.is.regExp], 'withText', '"text" argument', text);

        const apiFn = prepareApiFnArgs('withText', text);

        text = ensureRegExpContext(text);

        const selectorFn = () => {
            /* eslint-disable no-undef */
            const nodes = selector();

            if (!nodes.length) return null;

            return filterNodes(nodes, filter, document, void 0, textRe);
            /* eslint-enable no-undef */
        };

        const args = getDerivativeSelectorArgs(options, selectorFn, apiFn, filterByText, { textRe: (0, _makeRegExp2.default)(text) });

        return createDerivativeSelectorWithFilter(args);
    };

    obj.withExactText = text => {
        (0, _typeAssertions.assertType)(_typeAssertions.is.string, 'withExactText', '"text" argument', text);

        const selectorFn = () => {
            /* eslint-disable no-undef */
            const nodes = selector();

            if (!nodes.length) return null;

            return filterNodes(nodes, filter, document, void 0, exactText);
            /* eslint-enable no-undef */
        };

        const apiFn = prepareApiFnArgs('withExactText', text);
        const args = getDerivativeSelectorArgs(options, selectorFn, apiFn, filterByText, { exactText: text });

        return createDerivativeSelectorWithFilter(args);
    };

    obj.withAttribute = (attrName, attrValue) => {
        (0, _typeAssertions.assertType)([_typeAssertions.is.string, _typeAssertions.is.regExp], 'withAttribute', '"attrName" argument', attrName);

        const apiFn = prepareApiFnArgs('withAttribute', attrName, attrValue);

        attrName = ensureRegExpContext(attrName);

        if (attrValue !== void 0) {
            (0, _typeAssertions.assertType)([_typeAssertions.is.string, _typeAssertions.is.regExp], 'withAttribute', '"attrValue" argument', attrValue);
            attrValue = ensureRegExpContext(attrValue);
        }

        const selectorFn = () => {
            /* eslint-disable no-undef */
            const nodes = selector();

            if (!nodes.length) return null;

            return filterNodes(nodes, filter, document, void 0, attrName, attrValue);
            /* eslint-enable no-undef */
        };

        const args = getDerivativeSelectorArgs(options, selectorFn, apiFn, filterByAttr, {
            attrName,
            attrValue
        });

        return createDerivativeSelectorWithFilter(args);
    };

    obj.filter = (filter, dependencies) => {
        (0, _typeAssertions.assertType)([_typeAssertions.is.string, _typeAssertions.is.function], 'filter', '"filter" argument', filter);

        const apiFn = prepareApiFnArgs('filter', filter);

        filter = convertFilterToClientFunctionIfNecessary('filter', filter, dependencies);

        const selectorFn = () => {
            /* eslint-disable no-undef */
            const nodes = selector();

            if (!nodes.length) return null;

            return filterNodes(nodes, filter, document, void 0);
            /* eslint-enable no-undef */
        };

        const args = getDerivativeSelectorArgs(options, selectorFn, apiFn, filter);

        return createDerivativeSelectorWithFilter(args);
    };

    obj.filterVisible = () => {
        const apiFn = prepareApiFnArgs('filterVisible');
        const builder = new SelectorBuilder(getSelector(), { filterVisible: true, apiFn }, { instantiation: 'Selector' });

        return builder.getFunction();
    };

    obj.filterHidden = () => {
        const apiFn = prepareApiFnArgs('filterHidden');
        const builder = new SelectorBuilder(getSelector(), { filterHidden: true, apiFn }, { instantiation: 'Selector' });

        return builder.getFunction();
    };
}

function addCustomDOMPropertiesMethod({ obj, getSelector, SelectorBuilder }) {
    obj.addCustomDOMProperties = customDOMProperties => {
        assertAddCustomDOMPropertiesOptions(customDOMProperties);

        const builder = new SelectorBuilder(getSelector(), { customDOMProperties }, { instantiation: 'Selector' });

        return builder.getFunction();
    };
}

function addCustomMethodsMethod({ obj, getSelector, SelectorBuilder }) {
    obj.addCustomMethods = function (methods, opts) {
        assertAddCustomMethods(methods, opts);

        const customMethods = {};

        (0, _keys2.default)(methods).forEach(methodName => {
            customMethods[methodName] = {
                method: methods[methodName],
                returnDOMNodes: opts && !!opts.returnDOMNodes
            };
        });

        const builder = new SelectorBuilder(getSelector(), { customMethods }, { instantiation: 'Selector' });

        return builder.getFunction();
    };
}

function addHierarchicalSelectors(options) {
    const obj = options.obj;

    // Find

    obj.find = (filter, dependencies) => {
        (0, _typeAssertions.assertType)([_typeAssertions.is.string, _typeAssertions.is.function], 'find', '"filter" argument', filter);

        const apiFn = prepareApiFnArgs('find', filter);

        filter = convertFilterToClientFunctionIfNecessary('find', filter, dependencies);

        const selectorFn = () => {
            /* eslint-disable no-undef */
            return expandSelectorResults(selector, node => {
                if (typeof filter === 'string') {
                    return typeof node.querySelectorAll === 'function' ? node.querySelectorAll(filter) : null;
                }

                const results = [];

                const visitNode = currentNode => {
                    const cnLength = currentNode.childNodes.length;

                    for (let i = 0; i < cnLength; i++) {
                        const child = currentNode.childNodes[i];

                        results.push(child);

                        visitNode(child);
                    }
                };

                visitNode(node);

                return filterNodes(results, filter, null, node);
            });
            /* eslint-enable no-undef */
        };

        const args = getDerivativeSelectorArgs(options, selectorFn, apiFn, filter, { expandSelectorResults });

        return createDerivativeSelectorWithFilter(args);
    };

    // Parent
    obj.parent = (filter, dependencies) => {
        if (filter !== void 0) (0, _typeAssertions.assertType)([_typeAssertions.is.string, _typeAssertions.is.function, _typeAssertions.is.number], 'parent', '"filter" argument', filter);

        const apiFn = prepareApiFnArgs('parent', filter);

        filter = convertFilterToClientFunctionIfNecessary('find', filter, dependencies);

        const selectorFn = () => {
            /* eslint-disable no-undef */
            return expandSelectorResults(selector, node => {
                const parents = [];

                for (let parent = node.parentNode; parent; parent = parent.parentNode) parents.push(parent);

                return filter !== void 0 ? filterNodes(parents, filter, document, node) : parents;
            });
            /* eslint-enable no-undef */
        };

        const args = getDerivativeSelectorArgs(options, selectorFn, apiFn, filter, { expandSelectorResults });

        return createDerivativeSelectorWithFilter(args);
    };

    // Child
    obj.child = (filter, dependencies) => {
        if (filter !== void 0) (0, _typeAssertions.assertType)([_typeAssertions.is.string, _typeAssertions.is.function, _typeAssertions.is.number], 'child', '"filter" argument', filter);

        const apiFn = prepareApiFnArgs('child', filter);

        filter = convertFilterToClientFunctionIfNecessary('find', filter, dependencies);

        const selectorFn = () => {
            /* eslint-disable no-undef */
            return expandSelectorResults(selector, node => {
                const childElements = [];
                const cnLength = node.childNodes.length;

                for (let i = 0; i < cnLength; i++) {
                    const child = node.childNodes[i];

                    if (child.nodeType === 1) childElements.push(child);
                }

                return filter !== void 0 ? filterNodes(childElements, filter, node, node) : childElements;
            });
            /* eslint-enable no-undef */
        };

        const args = getDerivativeSelectorArgs(options, selectorFn, apiFn, filter, { expandSelectorResults });

        return createDerivativeSelectorWithFilter(args);
    };

    // Sibling
    obj.sibling = (filter, dependencies) => {
        if (filter !== void 0) (0, _typeAssertions.assertType)([_typeAssertions.is.string, _typeAssertions.is.function, _typeAssertions.is.number], 'sibling', '"filter" argument', filter);

        const apiFn = prepareApiFnArgs('sibling', filter);

        filter = convertFilterToClientFunctionIfNecessary('find', filter, dependencies);

        const selectorFn = () => {
            /* eslint-disable no-undef */
            return expandSelectorResults(selector, node => {
                const parent = node.parentNode;

                if (!parent) return null;

                const siblings = [];
                const cnLength = parent.childNodes.length;

                for (let i = 0; i < cnLength; i++) {
                    const child = parent.childNodes[i];

                    if (child.nodeType === 1 && child !== node) siblings.push(child);
                }

                return filter !== void 0 ? filterNodes(siblings, filter, parent, node) : siblings;
            });
            /* eslint-enable no-undef */
        };

        const args = getDerivativeSelectorArgs(options, selectorFn, apiFn, filter, { expandSelectorResults });

        return createDerivativeSelectorWithFilter(args);
    };

    // Next sibling
    obj.nextSibling = (filter, dependencies) => {
        if (filter !== void 0) (0, _typeAssertions.assertType)([_typeAssertions.is.string, _typeAssertions.is.function, _typeAssertions.is.number], 'nextSibling', '"filter" argument', filter);

        const apiFn = prepareApiFnArgs('nextSibling', filter);

        filter = convertFilterToClientFunctionIfNecessary('find', filter, dependencies);

        const selectorFn = () => {
            /* eslint-disable no-undef */
            return expandSelectorResults(selector, node => {
                const parent = node.parentNode;

                if (!parent) return null;

                const siblings = [];
                const cnLength = parent.childNodes.length;
                let afterNode = false;

                for (let i = 0; i < cnLength; i++) {
                    const child = parent.childNodes[i];

                    if (child === node) afterNode = true;else if (afterNode && child.nodeType === 1) siblings.push(child);
                }

                return filter !== void 0 ? filterNodes(siblings, filter, parent, node) : siblings;
            });
            /* eslint-enable no-undef */
        };

        const args = getDerivativeSelectorArgs(options, selectorFn, apiFn, filter, { expandSelectorResults });

        return createDerivativeSelectorWithFilter(args);
    };

    // Prev sibling
    obj.prevSibling = (filter, dependencies) => {
        if (filter !== void 0) (0, _typeAssertions.assertType)([_typeAssertions.is.string, _typeAssertions.is.function, _typeAssertions.is.number], 'prevSibling', '"filter" argument', filter);

        const apiFn = prepareApiFnArgs('prevSibling', filter);

        filter = convertFilterToClientFunctionIfNecessary('find', filter, dependencies);

        const selectorFn = () => {
            /* eslint-disable no-undef */
            return expandSelectorResults(selector, node => {
                const parent = node.parentNode;

                if (!parent) return null;

                const siblings = [];
                const cnLength = parent.childNodes.length;

                for (let i = 0; i < cnLength; i++) {
                    const child = parent.childNodes[i];

                    if (child === node) break;

                    if (child.nodeType === 1) siblings.push(child);
                }

                return filter !== void 0 ? filterNodes(siblings, filter, parent, node) : siblings;
            });
            /* eslint-enable no-undef */
        };

        const args = getDerivativeSelectorArgs(options, selectorFn, apiFn, filter, { expandSelectorResults });

        return createDerivativeSelectorWithFilter(args);
    };
}

function addAPI(selector, getSelector, SelectorBuilder, customDOMProperties, customMethods) {
    const options = { obj: selector, getSelector, SelectorBuilder, customDOMProperties, customMethods };

    addFilterMethods(options);
    addHierarchicalSelectors(options);
    addSnapshotPropertyShorthands(options);
    addCustomDOMPropertiesMethod(options);
    addCustomMethodsMethod(options);
    addCounterProperties(options);
}
