// testcafe/lib/client-functions/replicator.js
'use strict';

exports.__esModule = true;
exports.SelectorNodeTransform = exports.FunctionTransform = undefined;
exports.createReplicator = createReplicator;

var _lodash = require('lodash');

var _replicator = require('replicator');

var _replicator2 = _interopRequireDefault(_replicator);

var _builderSymbol = require('./builder-symbol');

var _builderSymbol2 = _interopRequireDefault(_builderSymbol);

var _compileClientFunction = require('../compiler/compile-client-function');

var _compileClientFunction2 = _interopRequireDefault(_compileClientFunction);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function createReplicator(transforms) {
    // NOTE: we will serialize replicator results
    // to JSON with a command or command result.
    // Therefore there is no need to do additional job here,
    // so we use identity functions for serialization.
    const replicator = new _replicator2.default({
        serialize: _lodash.identity,
        deserialize: _lodash.identity
    });

    return replicator.addTransforms(transforms);
}

// Replicator transforms
class FunctionTransform {
    constructor(callsiteNames) {
        this.type = 'Function';
        this.callsiteNames = callsiteNames;
    }

    shouldTransform(type) {
        return type === 'function';
    }

    toSerializable(fn) {
        const clientFnBuilder = fn[_builderSymbol2.default];

        if (clientFnBuilder) {
            return {
                fnCode: clientFnBuilder.compiledFnCode,
                dependencies: clientFnBuilder.getFunctionDependencies()
            };
        }

        return {
            fnCode: (0, _compileClientFunction2.default)(fn.toString(), null, this.callsiteNames.instantiation, this.callsiteNames.execution),
            dependencies: {}
        };
    }

    fromSerializable() {
        return void 0;
    }
}

exports.FunctionTransform = FunctionTransform;
class SelectorNodeTransform {
    constructor() {
        this.type = 'Node';
    }

    shouldTransform() {
        return false;
    }

    fromSerializable(nodeSnapshot) {
        return nodeSnapshot;
    }
}
exports.SelectorNodeTransform = SelectorNodeTransform;
