// testcafe/lib/load-assets.js
'use strict';

exports.__esModule = true;

exports.default = function (developmentMode) {
    const scriptNameSuffix = developmentMode ? 'js' : 'min.js';

    return {
        favIcon: loadAsset('./client/ui/favicon.ico', true),
        coreScript: loadAsset(`./client/core/index.${scriptNameSuffix}`),
        driverScript: loadAsset(`./client/driver/index.${scriptNameSuffix}`),
        uiScript: loadAsset(`./client/ui/index.${scriptNameSuffix}`),
        uiStyle: loadAsset('./client/ui/styles.css'),
        uiSprite: loadAsset('./client/ui/sprite.png', true),
        automationScript: loadAsset(`./client/automation/index.${scriptNameSuffix}`),

        // NOTE: Load the legacy client script lazily to reduce startup time
        legacyRunnerScript: require('testcafe-legacy-api').CLIENT_RUNNER_SCRIPT
    };
};

var _readFileRelative = require('read-file-relative');

const ASSETS_CACHE = {};

function loadAsset(filename, asBuffer) {
    if (!ASSETS_CACHE[filename]) ASSETS_CACHE[filename] = (0, _readFileRelative.readSync)(filename, asBuffer);

    return ASSETS_CACHE[filename];
}

module.exports = exports['default'];
